rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+        let bytes: Vec<u8> = base64::decode_config(b64bytes, base64::URL_SAFE_NO_PAD).unwrap();
+        let group = Key::group();
+        if bytes.len() != 65 || bytes[0] != 4 {
+            // It's not a properly tagged key.
+            return Err("Invalid key format for public key.".into());
+        }
+
+        let x = openssl::bn::BigNum::from_slice(&bytes[1..33]).expect("Invalid public key");
+        let y = openssl::bn::BigNum::from_slice(&bytes[33..]).expect("Invalid public key");
+        let mut new_key = match ec::EcKeyBuilder::new() {
+            Ok(key) => key,
+            Err(err) => return Err(format!("Could not generate key: {}", err)),
+        };
+        new_key.set_group(&group).expect("Could not set group");
+        new_key.set_public_key_affine_coordinates(&x, &y).expect(
+            "Invalid coordiates for public key",
+        );
+        Ok(Key(new_key.build()))
+    }
+}
+
+#[derive(Debug)]
+struct AuthElements {
+    t: Vec<String>,
+    k: String,
+}
+
+/// Parse the Authorization Header for useful things.
+fn parse_auth_token(auth_token: &mut String) -> Result<AuthElements, String> {
+    let mut parts: Vec<&str> = auth_token.split(" ").collect();
+    let mut schema = parts.remove(0).to_lowercase();
+    // Ignore the first token if it's the header line.
+    if schema == "authorization:" {
+        schema = parts.remove(0).to_lowercase();
+    }
+    let mut reply: AuthElements = AuthElements {
+        t: Vec::new(),
+        k: String::from(""),
+    };
+    match schema.to_lowercase().as_ref() {
+        "vapid" => {
+            let sub_parts: Vec<&str> = parts[0].splitn(2, ",").collect();
+            for kvi in &sub_parts {
+                let kv: Vec<String> = kvi.splitn(2, "=").map(|x| String::from(x)).collect();
+                match kv[0].to_lowercase().as_ref() {
+                    "t" => {
+                        let ts: Vec<String> = kv[1].split(".").map(|x| String::from(x)).collect();
+                        if ts.len() != 3 {
+                            return Err("Invalid t token specified".into());
+                        }
+                        let ttoken = format!("{}.{}", ts[0], ts[1]);
+                        reply.t = vec![ttoken, ts[2].clone()];
+                    }
+                    "k" => reply.k = String::from(kv[1].clone()),
+                    _ => {}
+                }
+            }
+        }
+        "webpush" => {
+            reply.t = parts[0].split(".").map(|x| String::from(x)).collect();
+        }
+        _ => return Err(format!("Unknown schema type: {}", parts[0])),
+    };
+    return Ok(reply);
+}
+
+// Preferred schema
+static SCHEMA: &str = "vapid";
+
+/// Convert the HashMap containing the claims into an Authorization header.
+/// `key` must be generated or initialized before this is used. See `Key::from_pem()` or
+/// `Key::generate()`.
+pub fn sign(key: Key, claims: &mut HashMap<String, serde_json::Value>) -> Result<String, String> {
```

Comment:
If you'd like you can typically change types like `HashMap<String, serde_json::Value>` into a more "rustic" datatype using the various serde attributes to control serialization. For example you could have something like:

```rust
#[derive(Serialize, Deserialize)]
struct Claims {
    #[serde(skip_serializing_if = "Option::is_none")]
    sub: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    exp: Option<u64>,
}
```

(or stuff like that)

It in general just makes it a little easier to work with in Rust and provides some more input/output validation for later on.

Reply:
Oh, I should look more into that. Apparently, I just looked at the serde_json library. Thanks!